src/catnip/runtime/memory/mempool.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//==============================================================================
+// Imports
+//==============================================================================
+
+use ::dpdk_rs::{
+    rte_mbuf,
+    rte_mempool,
+    rte_pktmbuf_alloc,
+    rte_pktmbuf_clone,
+    rte_pktmbuf_free,
+    rte_pktmbuf_pool_create,
+    rte_socket_id,
+};
+use ::runtime::fail::Fail;
+use ::std::ffi::CString;
+
+//==============================================================================
+// Structures
+//==============================================================================
+
+/// DPDK Memory Pool
+#[derive(Debug)]
+pub struct MemoryPool {
+    /// Underlying memory pool.
+    pool: *mut rte_mempool,
+}
+
+//==============================================================================
+// Associate Functions
+//==============================================================================
+
+/// Associated functions for memory pool.
+impl MemoryPool {
+    /// Creates a new memory pool.
+    pub fn new(
+        name: CString,
+        data_room_size: usize,
+        pool_size: usize,
+        cache_size: usize,
+    ) -> Result<Self, Fail> {
+        let pool: *mut rte_mempool = unsafe {
+            rte_pktmbuf_pool_create(
+                name.as_ptr(),
+                pool_size as u32,
+                cache_size as u32,
+                0,
+                data_room_size as u16,
+                rte_socket_id() as i32,
+            )
+        };
+
+        // Failed to create memory pool.
+        if pool.is_null() {
+            return Err(Fail::new(libc::EAGAIN, "failed to create memory pool"));
+        }
+
+        Ok(Self { pool })
+    }
+
+    /// Gets a raw pointer to the underlying memory pool.
+    pub fn into_raw(&self) -> *mut rte_mempool {
+        self.pool
+    }
+
+    /// Allocates a mbuf in the target memory pool.
+    pub fn alloc_mbuf(&self, size: Option<usize>) -> Result<*mut rte_mbuf, Fail> {
+        // TODO: Drop the following warning once DPDK memory management is more stable.
+        warn!("allocating mbuf from DPDK pool");
+
+        // Allocated mbuf.
+        let mut mbuf_ptr: *mut rte_mbuf = unsafe { rte_pktmbuf_alloc(self.pool) };
+        if mbuf_ptr.is_null() {
+            return Err(Fail::new(libc::ENOMEM, "cannot allocate more mbufs"));
+        }
+
+        // Fill out some fields of the underlying mbuf.
+        unsafe {
+            let mut num_bytes: u16 = (*mbuf_ptr).buf_len - (*mbuf_ptr).data_off;
+
+            if let Some(size) = size {
+                // Allocated buffer is not  big enough.
```

Comment:
Comments about an if that come before the if should describe what it is checking, not one of the branches.  Descriptions of one of the branches should go inside that branch of the if.

For example:
// Check whether x is big enough.
if x > 5 {
    // x is big enough to do something.
    do_something();
} else {
    // x is too small.
    do_something_else();
}

To me, what you have on line 84 is describing the result of the if on line 83.  But you actually mean to describe the result of the if on line 85.  I'd combine the line 84 comment with the comment on line 86, and say "The allocated buffer is not big enough, rollback the allocation."  If you want a comment on line 84, it should say something like "Check if the allocated buffer is big enough."

Reply:
Basically, the comment should pertain to the code block in which it resides.

Reply:
Sure. I'll make it like you suggested.